clang/test/CodeGenSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,138 @@
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[MACRO_MACRO_SIZE]], [[MACRO_MACRO_SIZE]]* @[[MACRO_MACRO_Y]]
+
+        template_param<int>();
+        // CHECK: define linkonce_odr spir_func void @_Z14template_paramIiEvv
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[INT_SIZE]], [[INT_SIZE]]* @[[INT]]
+
+        template_param<decltype(x)>();
+        // CHECK: define internal spir_func void @"_Z14template_paramIZZ4mainENK3$_0clEvEUlvE_Evv"
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[LAMBDA_SIZE]], [[LAMBDA_SIZE]]* @[[LAMBDA]]
+
+        lambda_in_dependent_function<int>();
+        // CHECK: define linkonce_odr spir_func void @_Z28lambda_in_dependent_functionIiEvv
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[DEP_INT_SIZE]], [[DEP_INT_SIZE]]* @[[LAMBDA_IN_DEP_INT]]
+
+        lambda_in_dependent_function<decltype(x)>();
+        // CHECK: define internal spir_func void @"_Z28lambda_in_dependent_functionIZZ4mainENK3$_0clEvEUlvE_Evv"
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[DEP_LAMBDA_SIZE]], [[DEP_LAMBDA_SIZE]]* @[[LAMBDA_IN_DEP_X]]
+
+        lambda_no_dep<int, double>(3, 5.5);
+        // CHECK: define linkonce_odr spir_func void @_Z13lambda_no_depIidEvT_T0_(i32 %a, double %b)
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[NO_DEP_LAMBDA_SIZE]], [[NO_DEP_LAMBDA_SIZE]]* @[[LAMBDA_NO_DEP]]
+
+        int a = 5;
+        double b = 10.7;
+        auto y = [](int a) { return a; };
+        auto z = [](double b) { return b; };
+        lambda_two_dep<decltype(y), decltype(z)>();
+        // CHECK: define internal spir_func void @"_Z14lambda_two_depIZZ4mainENK3$_0clEvEUliE_ZZ4mainENKS0_clEvEUldE_Evv"
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[DEP_LAMBDA1_SIZE]], [[DEP_LAMBDA1_SIZE]]* @[[LAMBDA_TWO_DEP]]
+
+        lambda_two_dep<decltype(z), decltype(y)>();
+        // CHECK: define internal spir_func void @"_Z14lambda_two_depIZZ4mainENK3$_0clEvEUldE_ZZ4mainENKS0_clEvEUliE_Evv"()
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[DEP_LAMBDA2_SIZE]], [[DEP_LAMBDA2_SIZE]]* @[[LAMBDA_TWO_DEP2]]
+      });
+
+  kernel_single_task<class kernel2>(func<Derp>);
+  // CHECK: define internal spir_func void @_Z18kernel_single_taskIZ4mainE7kernel2PFPKcvEEvT0_
+  // CHECK: call spir_func void @_Z18kernel_single_taskIZ4mainE7kernel2PFPKcvEEvT0_(i8* ()* @_Z4funcI4DerpEDTu20__unique_stable_nameXsrT_3strEEEv)
+
+  auto l1 = []() { return 1; };
+  auto l2 = [](decltype(l1) *l = nullptr) { return 2; };
+  kernel_single_task<class kernel3>(l2);
+  printf(__builtin_unique_stable_name(l2));
+  // CHECK: define internal spir_func void @"_Z18kernel_single_taskIZ4mainE7kernel3Z4mainE3$_1EvT0_"
+  // CHECK: call spir_func i32 @"_ZZ4mainENK3$_1clEPZ4mainE3$_2"
+}
```

Comment:
Missing a codegen test for the VLA behavior discussed in previous reviews?

Reply:
VLA test 2 below  
```
 int j = 55;
  printf(__builtin_unique_stable_name(int[j])); // No warning
    // Ensure that j is incremented, because VLAs are terrible 
```

unable to see any increment instruction in the emitted IR (shown below), however,  
`_ZTSAppL_ZZ4mainE1jE_i` demangles to `typeinfo name for int [(main::j)++]`


[] clang> clang -cc1 -triple spir64-unknown-unknown-sycldevice -fsycl-is-device -disable-llvm-passes -emit-llvm -o - test
/CodeGenSYCL/unique_stable_name.cpp
```
; ModuleID = 'test/CodeGenSYCL/unique_stable_name.cpp'
source_filename = "test/CodeGenSYCL/unique_stable_name.cpp"
target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024"
target triple = "spir64-unknown-unknown-sycldevice"
 @usn_str = private unnamed_addr constant [23 x i8] c"_ZTSAppL_ZZ4mainE1jE_i\00", align 1

; Function Attrs: convergent noinline norecurse nounwind optnone mustprogress
define dso_local spir_func void @printf(i8* %0) #0 {
entry:
  %.addr = alloca i8*, align 8
  store i8* %0, i8** %.addr, align 8
  ret void
}

; Function Attrs: convergent noinline norecurse nounwind optnone mustprogress
define dso_local i32 @main() #0 {
entry:
  %j = alloca i32, align 4
  store i32 55, i32* %j, align 4
  call spir_func void @printf(i8* getelementptr inbounds ([23 x i8], [23 x i8]* @usn_str, i32 0, i32 0)) #1
  ret i32 0
}

attributes #0 = { convergent noinline norecurse nounwind optnone mustprogress "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" }
attributes #1 = { convergent }

!llvm.module.flags = !{!0}
!llvm.ident = !{!1}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{!"clang version 13.0.0 (https://github.com/erichkeane/llvm-project e5ac8fce4eb047f9e2cb0fdf893a05235eee3a71)"}

```

Reply:
I think we didn't end up implementing that side-effect, the code that does so is VERY specific about being for sizeof.  Initially we intended to limit to only complete sized types to avoid that, but we weren't able to due to some needs of SYCL.

I also think that we miss the warning for the same reason.  IMO, we'd probably be ok with a `// FIXME` in the test for those two situations.  WDYT @AaronBallman ?

Reply:
Added test cases with FIXMEs